lib/spack/spack/package_base.py

```diff
@@ -1037,8 +1039,46 @@ def resolve_binary_provenance(self) -> None:
         Base implementation will look up git commits when appropriate.
         Packages may override this implementation for custom implementations
         """
-        # TODO in follow on PR adding here so SNL team can begin work ahead of spack core
-        pass
+
+        # early return cases, don't overwrite user intention
+        # commit pre-assigned or develop specs don't need commits changed
+        # since this would create un-necessary churn
+        if "commit" in self.spec.variants or self.spec.is_develop:
+            return
+
+        sha = None
+
+        if is_git_version(str(self.spec.version)):
+            ref = self.spec.version.ref
+        else:
+            tag = self.version_or_package_attr("tag", self.spec.version, None)
+            branch = self.version_or_package_attr("branch", self.spec.version, None)
+            assert not (tag and branch)
+            ref = tag or branch
+
+        assert ref, "Missing git ref"
+
+        if not sha:
+            url = self.version_or_package_attr("git", self.spec.version)
+            sha = spack.util.git.get_commit_sha(url, ref)
+
+        # we typically check mirrors first, but for commit resolution we query the network
+        # and fall back to mirrors assuming people want the most up-to-date commits
+        if not sha:
+            if self.stage.expanded:
+                sha = spack.util.git.get_commit_sha(self.stage.source_path, ref)
+            if not sha:
+                try:
+                    self.do_fetch(mirror_only=True)
+                except spack.error.FetchError:
+                    pass
+                if self.stage.archive_file:
+                    sha = spack.util.archive.retrieve_commit_from_archive(
+                        self.stage.archive_file, ref
+                    )
```

Comment:
This whole block can be dedented or turned into an else-block on the previous condition -- I guess the latter is probably preferable?